README.md

# geoMatch

> geoMatch

A package designed to provide matched data for propensity-score based causal analyses using spatially-explicit data.  Fully compatible with matching strategies available in the MatchIt package, geoMatch provides adjusted outcome measures which mitigate bias associated with spillover from areas which received interventions to proximate areas that did not.

## Installation
install.packages("geoMatch")
geoMatch imports both the sp and MatchIt packages.

## Usage

```r
#A full demo for usage can be found in /demo/match.spatial.data.R
library(geoMatch)
##Simulate Latitude and Longtiude information for each point, 
##with enforced spatial correlation.
library(sp)
set.seed(500)
coords = cbind(runif(nrow(lalonde),37.1708,37.3708), 
               runif(nrow(lalonde),76.6069,76.8069))

##Create a spatial points data frame
spatial_lalonde <- SpatialPointsDataFrame(coords, lalonde)

match.out2 <- geoMatch(treat ~ age + educ + black + hispan + nodegree + 
                         married + re74 + re75, 
                      method = "nearest", 
                      caliper=0.25, 
                      data = spatial_lalonde, 
                      outcome.variable="re78", 
                      outcome.suffix="_adjusted")
lm.out2 <- lm(re78_adjusted ~ treat + age + educ + black + hispan + nodegree + 
                married + re74 + re75 + distance, 
              data = match.data(match.out2))
summary(lm.out2)
```

## License

Copyright 2016 Daniel Miller Runfola (dan@danrunfola.com)

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



NEWS.md
# 0.0.1
Internal alpha candidate release of the geoMatch package - Spatial Causal Inference (Sept 10 2014).

# 0.0.2
Devtools release for institutional testing (Feb 15 2015).

# 0.0.3
SciClone testing edition, improved hooks for cluster environments (Mar 15 2015).

# 0.0.3
Closer alignment with MatchIt functionality, new distance decay function availability (Nov 17 2015).

# 0.0.6
Improvements in the effeciency of distance decay calculations (Mar 28 2016).

# 0.0.7
Rewrite to implement spatial adjustment factors and optimizations for more flexible integration with MatchIt (Sep 19 2016).

# 0.1.0
Improved fitting effeciencies for cluster environment simulation testing (Oct 1 2016).

# 0.1.1
Release for CRAN consideration with updates to facilitate CRAN compatability (Oct 19 2016).


R/geoMatch.R
#' geoMatch
#'
#' geoMatch improves models using spatial data for the purposes of causal inference by selecting matched subsets of the original treated and control groups.
#' It provides an extension of the R package MatchIt (see Ho, Imai, King, and Stuart (2004)), enabling the use 
#' of spatial data frames as well as providing an adjustment factor to mitigate potential spillover between 
#' treated and control units (i.e., in cases where stable unit treatment value assumptions (SUTVA) may not be met
#' due to spatial spillovers).  geoMatch maintains the full functionality of MatchIt, including all 
#' matching strategies and caliper functions. Full documentation on MatchIt is available online at http://gking.harvard.edu/matchit, 
#' and help for specific commands is available through help.matchit.
#'
#'
#' @docType package
#' @name geoMatch
#' @export
#' @import sp
#' @import MatchIt
#' @details
#' geoMatch overcomes two challenges.  First, it seeks to first remove spillover from control outcome measures,
#' for example a case where a clinic may improve health outcomes in both the geographic neighborhood
#' it is located in, as well as nearby neighborhoods.  Failing to adjust for this spillover can result in
#' erroneous estimates of impact. Second, it allows for the use of spatial points data frames in conjunction
#' with the MatchIt matching framework.
#' geoMatch returns an adjusted version of the outcome variable for each control case as specified by the user, 
#' defined as Y*.  Y* can be interpreted as the estimated outcome if the spatial spillovers from treated
#' cases are netted out. This adjusted Y* can be used in the second stage model for more accurate estimations of
#' treatment effects.  Y* is calculated through a multiple step process.
#' First, a distance matrix (Dct) is constructed which provides the Euclidean distances between each treated and control case:
#' \preformatted{
#' [Dc1t1  Dc1t2   Dc1t3  .   Dc1tn]
#' [  .      .       .    .     .  ]
#' [  .      .       .    .     .  ]
#' [Dcnt1  Dcnt2   Dcnt3  .   Dcntn]
#' }
#' Second, two vector (Yt and Yc) are constructed which contain the known, observed outcomes in both Yt and Yc.
#' Third, a spherical distance decay function is fit for each T simultanesouly, in which the parameter Ut is 
#' solved for across all units T in the spatial function sf(distance, U):
#' \preformatted{
#' (1 - [[3/2] * (Dct / Ut) - [1/2] * (Dct/Ut)^3])
#' }
#' where we seek to optimize the absolute difference between the observed outcome at each control location (Yc) and the estimated
#' outcome (Yc_e) as a product of neighboring units Yt:
#' \preformatted{
#' Yc_e =  sf(Dct, Ut) * Yt 
#' minimize(abs(Yc_e-Yc))
#' }
#' The vector of distances Ut is used to estimate the 
#' adjusted Yc*, which - for each control case - provides an estimate of the outcome
#' if spillover is removed:
#' \preformatted{
#' Yc* = Yc - (sf[Dct, Ut]*Yt).
#' }
#' geoMatch then proceeds as usual with MatchIt, returning a matched set of spatial locations alongside Yc*.
#'@param ... The first parameters provided to geoMatch should be a traditional MatchIt specification.  Full documentation on MatchIt is available online at http://gking.harvard.edu/matchit. Dataframe must be a spatial points dataframe.
#'@param outcome.variable The name of the outcome variable that will be modeled to establish causal effect.  This must be an existing attribute in the spatial dataframe passed to geoMatch.
#'@param outcome.suffix Suffix for the returned column name with spillover-adjusted outcome data.
#'@return This function will return a MatchIt object, with a spatial data frame accesible in $spdf.
#'@example /demo/match.spatial.data.R
#'@source \url{http://geo.aiddata.org/}
#'@section Authors:
#'Dan Miller Runfola dan@danrunfola.com; Ariel BenYishay abenyishay@wm.edu; Seth Goodman sgoodman@aiddata.org
#'@section Citation:
#' \preformatted{
#' If you find this package useful, please cite:
#'Runfola, D., BenYishay, A., Goodman, S., 2016. geoMatch: An R Package for Spatial Propensity Score Matching. R package. http://geo.aiddata.org.
#'Daniel Ho, Kosuke Imai, Gary King, and Elizabeth Stuart (2007). Matching as Nonparametric Preprocessing for Reducing Model Dependence in Parametric Causal Inference. Political Analysis 15(3): 199-236. http://gking.harvard.edu/files/abs/matchp-abs.shtml
#'}

geoMatch <- function (..., outcome.variable,outcome.suffix="_adjusted")
{
  spatial_match = -1
  a <- list(...)
  if (missing(outcome.variable) & isS4(a[['data']]))
  {
    warning("You are using a spatial dataframe, but did not assign a outcome.variable. This is required for spatial adjustments.  MatchIt will be performed ignoring spatial options.")
    spatial_match = FALSE
    if("data" %in% names(a))
    {
      #Pass the data name forward for a cleaner summary
      dName <- as.character(match.call()$data)
    }
    dfA <- as.data.frame(a[['data']]@data)
    a[['data']] <- dfA
  }
  
  if(!missing(outcome.variable) & !isS4(a[['data']]) & spatial_match==-1)
  {
    warning("You assigned an outcome variable, but your data is not a spatial data frame. A spatial data frame is required for spatial adjustments.  MatchIt will be performed ignoring spatial options.")
    spatial_match = FALSE  
    if("data" %in% names(a))
    {
      #Pass the data name forward for a cleaner summary
      dName <- as.character(match.call()$data)
    }
  }
  if(missing(outcome.variable) & !isS4(a[['data']]) & spatial_match==-1)
  {
    warning("To use the spatial functions of geoMatch, you must specify a spatial data frame and outcome.variable.  MatchIt is being run ignoring spatial options.")
    spatial_match = FALSE
    if("data" %in% names(a))
    {
      #Pass the data name forward for a cleaner summary
      dName <- as.character(match.call()$data)
    }
    
  }
  
  if(spatial_match == FALSE)
  {
    m.out <- do.call("matchit", a)
    m.out$call$data <- dName
    return(m.out)
  }
  
  if(!missing(outcome.variable) & isS4(a[['data']]) & spatial_match != FALSE)
  {
    x.coord <- coordinates(a[['data']])[,2]
    y.coord <- coordinates(a[['data']])[,1]
    
    if(max(x.coord) > 180.0 | min(x.coord) < -180.0 | max(y.coord) > 90.0 | min(y.coord) < -90.0)
    {
      warning("geoMatch currently only supports data projected with latitude and longitude information (WGS84).  Please reproject your data.")
      return(0)
    }
    
    if(class(a[['data']])[1] == "SpatialPointsDataFrame")
    {
      o_var <- outcome.variable
      geoMatch.Core(..., outcome.variable = o_var) 
    }
    else
    {
      warning("Currently only spatial points data frames are supported.  Please convert your spatial data to a Spatial Points Data Frame.")
      return(0)
    }
  }
  else
  {
    warning("geoMatch encountered an error. Ensure that you have specified a valid spatial dataframe and outcome.variable.")
    return(0)
  }
  
}





R/geoMatchCore.R
#Core functionality for the geoMatch MatchIt wrapper.
geoMatch.Core <- function (..., outcome.variable,outcome.suffix="_adjusted"){
  a <- list(...)
  
  #Remove spillover from all C outcomes, then matches.
  #Mitigates OVB through PSM matching, but does not take
  #advantage of OVB back-extrapolation.
  #To correct for potential OVB in spatially-correlated systems,
  #see geoOVB.
  
  #For each matched unit, this outputs a new variable (Y*), which is
  #the outcome with spatial spillovers from T netted out.
  #This should be used in your second stage model.
  
  #Dct - Distance between each control case and treatment case (matrix)
  #[Dc1t1  Dc1t2   Dc1t3  .   Dc1tn]
  #[  .      .       .    .     .  ]
  #[  .      .       .    .     .  ]
  #[Dcnt1  Dcnt2   Dcnt3  .   Dcntn]


  
  t.name <- strsplit(as.character(a[[1]]),"~")[[2]]
  t.coords <- a[['data']][as.vector(a[['data']]@data[t.name]==1),]
  c.coords <- a[['data']][as.vector(a[['data']]@data[t.name]==0),]
  Dct <- matrix(, nrow = length(c.coords), ncol = length(t.coords))
  rownames(Dct) <- rownames(c.coords@data)
  colnames(Dct) <- rownames(t.coords@data)
  for(i in 1:(length(c.coords)))
  {
    ct <- spDistsN1(t.coords, c.coords[i,], longlat = TRUE)
    Dct[i,] <- ct
  }
  
  #Yt - Vector. Known outcome at each T.
  Yt.gen <- a[['data']][as.vector(a[['data']]@data[t.name]==1),]
  Yt <- Yt.gen@data[outcome.variable]
  
  #Yc - Vector. Known outcome at each C.
  Yc.gen <- a[['data']][as.vector(a[['data']]@data[t.name]==0),]
  Yc <- Yc.gen@data[outcome.variable]
  
  #e - residual variance for units c and/or t (depending on subscript)
  #sf(Dct, Ut) - A spatial decay function, variably defined.
  #This iteration of the script assumes spherical:
  #(1 - [[3/2] * (Dct / Ut) - [1/2] * (Dct/Ut)^3])
  
  #Ut - Vector. Parameter "U" in spherical 
  #distance decay relationships, solved for 
  #each T simultaneously
  
  #Solve for Ut in Yc =  sf(Dct, Ut) * Yt + ec
  

  sf.opt <- function(Ut, ...)
  {
    Yc.spill.est.genA = ((3/2) * (Dct / Ut) - (1/2) * (Dct/Ut)^3)
    Yc.spill.est.genA[Yc.spill.est.genA < 0.0] <- 0
    Yc.spill.est.genB <- sweep(Yc.spill.est.genA,MARGIN=2,Yt[[1]],'*')
    Yc.spill.est <- rowSums(Yc.spill.est.genB)
    Yc.err = sum(abs(Yc - Yc.spill.est))
    #Ineffecient approach to preventing negative distances
    return(Yc.err)
  }
  
  sf <- function(...)
  {
    Yc.spill.est.genA = ((3/2) * (Dct / Ut) - (1/2) * (Dct/Ut)^3)
    Yc.spill.est.genA[Yc.spill.est.genA < 0.0] <- 0
    Yc.spill.est.genB <- sweep(Yc.spill.est.genA,MARGIN=2,Yt[[1]],'*')
    Yc.spill.est <- rowSums(Yc.spill.est.genB)
    return(Yc.spill.est)
  }
  

  #Optimize Ut vector, limiting distance threshold to be
  #greater than 0, and less than the circumfrence of the earth 
  #(Approx. 40,100 km)
  #Use random starting points between the minimum and maximum observed distances
  #between C and T.
  Ut <- runif(nrow(Yt),(min(Dct)+.00001),max(Dct))
  #print(Ut)
  m_init <- round(max(Dct)*4,0)
  Ut.optim <- 
    optimx(par = Ut, 
        fn=sf.opt, 
        gr=NULL,
        hess=NULL,
        method = "spg",
        lower = 0.01,
        upper=m_init,
        itnmax=100000,
        #hessian=FALSE,
        control=list(trace=0),
        Dct)
  
  if(Ut.optim$convcode != 0)
  {
    warning("No optimal spatial decay function was found, and the below message (if any) was returned by the optimization procedure.  geoMatch will attempt to continue with the best available parameters.")
    warning(Ut.optim$message)
  }
  

    Ut <- Ut.optim[1:length(Ut)]


  #Calculate adjusted Yc*, which - for each C - removes spatial spillover.
  #Yc* = Yc - (sf[Dct, Ut]*Yt) [Note: Yt multiplier is applied in the function
  #to make this code easier to read].
  
  Yc.star <- Yc - sf(Dct, Ut)
  spillovers_c <- sf(Dct, Ut)

  
  #Update the original outcome values to remove spillover.
  outcome.variable.adjusted <- paste(outcome.variable,outcome.suffix,sep="")
  a[['data']]@data[outcome.variable.adjusted] <- a[['data']]@data[outcome.variable]
  a[['data']]@data[a[['data']]@data[t.name]==0,][outcome.variable.adjusted] <- Yc.star
  a[['data']]@data["est_spillovers"] <- 0
  a[['data']]@data[a[['data']]@data[t.name]==0,]["est_spillovers"] <- spillovers_c
  
  #Object with pre- and post statistics.
  pre.post.spatial <- 
    c(summary(a[['data']]@data[a[['data']]@data[t.name]==0,][outcome.variable]),
      summary(a[['data']]@data[a[['data']]@data[t.name]==0,][outcome.variable.adjusted]))
  
  #Run matchIt as usual for the user.
  spdfA <- a[['data']]
  dfA <- as.data.frame(a[['data']]@data)
  a[['data']] <- dfA
  m.res <- do.call("matchit", a)

  #Fix the data call to account for the do.match and (for now!) lack of substitute.
  m.res$call$data <- as.list(substitute(list(...)))[-1L]$data
  
  #Update the source data with the new column.
  #This is to play nice with MatchIt's match.data() call.
  
  env <- attributes(terms(...))$.Environment
  
  assign_to <- paste(m.res$call$data,"@data$",outcome.variable.adjusted,sep="")
  assign_dta <- a[['data']][outcome.variable.adjusted]
  
  c.call <- paste(assign_to, "<-", assign_dta)
  
  eval(parse(text=c.call), envir = env)
  assign(assign_to, assign_dta, envir=env, inherits=TRUE)

  
  spdfA$weights <- m.res$weights
  spdfA$matched <- m.res$weights>0
  spdfA$distance <- m.res$distance
  
  m.res$spdf <- spdfA
  
  return(m.res)
  
  #These adjusted values can then be used in a traditional matchIt.
  
  

}

geoMatch/R/package.R
#' geoMatch
#'
#' A package designed to provide matched data for propensity-score based causal analyses using spatially-explicit data
#'
#' @docType package
#' @name geoMatch

geoMatch <- function (match,mod,dta)
{
  #First, calculate the OVB function (third order poly for now).
  
  #Second, for every Treated unit
    #use MatchIt as specified by the user to identify the best match
    #(or set of matches).
    #Take the difference in outcomes from each T to the matched
    #C(s). If multiple Cs are matched, use MatchIt weights.
  
      #Repeat this over multiple distance thresholds (for every D,
      #Remove all Cs which are within the D threshold of ANY T, not just the
      #One being analyzed).
      #Record the distance (or, average dist) of the best found match from the T unit.
  
    #After all distance thresholds have been run
    #Parameterize a distance-decay model (3rd order poly for now)
    #that summarizes the estimates spillover for that TE (Si)
  
  #Third, run a normal, full MatchIt model according to the users specified MatchIt.
  
  #Fourth, Calculate the Distances between all control cases included in the model and the nearest T.
  
  #Fifth, adjust each C's outcome measure according to the estimated spillover, given
  #the distance to the closest T (assuming non-additive spillover) as well as the OVB function.
  
  #Sixth, this function returns the new dataframe with the distance-adjusted outcomes,
  #which do not include spillover effects, as well as the MatchIt pairs.
  
  #Outside of this function, the user runs their model with the returned, adjusted, 
  #matched and transformed dataset.

  
  
}


demo/match.spatial.data.R
library(geoMatch)

###
### An Example Script for Obtaining Matched Data when you have
### Spatial information
###
data(lalonde)

##Traditional, non-spatial MatchIt
library(MatchIt)
match.out1 <- matchit(treat ~ age + educ + black + hispan + 
                        nodegree + married + re74 + re75, 
                  method = "nearest", data = lalonde)

##Example model performed after matching, including both Control and Treatment groups
lm.out1 <- lm(re78 ~ treat + age + educ + black + hispan + 
                nodegree + married + re74 + re75 + distance, 
              data = match.data(match.out1))
summary(lm.out1)

##Simulate Latitude and Longtiude information for each point, 
##with enforced spatial correlation.
library(sp)
set.seed(500)
coords = cbind(runif(nrow(lalonde),37.1708,37.3708), 
               runif(nrow(lalonde),76.6069,76.8069))

##Create a spatial points data frame
spatial_lalonde <- SpatialPointsDataFrame(coords, lalonde)

##Matching and adjusting for spillover effects
##See ?geoMatch for more parameters specific to spatial data.
##See ?MatchIt for more options for matching methods.
match.out2 <- geoMatch(treat ~ age + educ + black + hispan + nodegree + 
                         married + re74 + re75, 
                      method = "nearest", 
                      caliper=0.25, 
                      data = spatial_lalonde, 
                      outcome.variable="re78", 
                      outcome.suffix="_adjusted",
                      optim.iterations = 100)


##Example maps 
spplot(match.out2$spdf, z="matched", col.regions=c("red","green"), 
       main="Map of Matched Pairs")
spplot(match.out2$spdf, z="distance", main="Propensity Scores")
spplot(match.out2$spdf, z="est_spillovers", main="Estimated Spillovers")
spplot(match.out2$spdf, z="re78_adjusted", main="Adjusted Outcome")

#Percent of outcomes attributable to spillovers
match.out2$spdf@data["spill_percent"] <- 100 * 
  (match.out2$spdf@data["est_spillovers"] / match.out2$spdf@data["re78"])

spplot(match.out2$spdf[!is.infinite(match.out2$spdf@data$spill_percent) & 
                         match.out2$spdf@data$treat == 0,], 
       z="spill_percent", 
       main="% Outcome Attributable to Spillover",
       pretty=TRUE,
       cuts=5)

##Example model performed after spatial spillover adjustment, using matched data
lm.out2 <- lm(re78_adjusted ~ treat + age + educ + black + hispan + nodegree + 
                married + re74 + re75 + distance, 
              data = match.data(match.out2))
summary(lm.out2)

##Example model with spatial lag after spatial spillover adjustment
##For more information on practical specifications for these models, see
##Corrado, Luisa, and Bernard Fingleton. Where is the economics in spatial 
##econometrics?. Journal of Regional Science 52.2 (2012): 210-239.
library(spdep)
matched.spatial <- match.out2$spdf[match.out2$spdf@data$matched == 1,]
coords <- coordinates(matched.spatial)
k1 <- knn2nb(knearneigh(coords, k=15))
all.linked <- max(unlist(nbdists(k1, coords)))
neighbor.list <- dnearneigh(coords, 0, all.linked)
sl.out3 <- lagsarlm(re78 ~ treat + age + educ + black + hispan + 
                      nodegree + married + re74 + re75 + distance, 
                    data=matched.spatial,
                        nb2listw(neighbor.list, style="W"), method="MC", quiet=TRUE,
                        tol.solve=1.0e-16)
summary(sl.out3)
